core/src/main/java/org/infinispan/reactive/publisher/impl/LocalPublisherManagerImpl.java

```diff
@@ -236,39 +237,58 @@ private void updateStrategy(boolean usingSegmentedStored) {
                                                                         FlowableConverter<K, Flowable<I>> conversionFunction,
                                                                         Function<? super Publisher<I>, ? extends Publisher<R>> transformer) {
       return new BaseSegmentAwarePublisherSupplier<R>() {
-         private Publisher<R> actualPublisherValues() {
+         @Override
+         public Publisher<R> publisherWithoutSegments() {
             return Flowable.fromIterable(keysToInclude)
                   .to(conversionFunction)
                   .to(transformer::apply);
          }
 
          @Override
-         public Publisher<R> publisherWithoutSegments() {
-            return actualPublisherValues();
-         }
-
-         @Override
-         Flowable<NotificationWithLost<R>> flowableWithNotifications() {
-            return Flowable.fromPublisher(actualPublisherValues())
-                  .map(Notifications::value)
+         Flowable<NotificationWithLost<R>> flowableWithNotifications(boolean reuseNotifications) {
+            return Flowable.fromIterable(keysToInclude)
+                  .groupBy(keyPartitioner::getSegment)
+                  // We use concatMapEager to complete segments in the iteration order to ensure
+                  // ordering guarantees defined in the LocalPublisherManager entryPublisher method
```

Comment:
Yeah, we have to use `concatMapEager` so that we subscribe to all of the grouped publishers at the same time even if we don't publish their values otherwise we can starve groupBy. This is different than `flatMap` in that the ordering of the groups are maintained. This way we can still guarantee we have the new guarantees defined by LocalPublisherManager that state each segment is fully retrieved before publishing a new one.

Reply:
Got it now, I thought the alternative was `concatMap`, not `flatMap`